Add java example

Taowyoo opened the pull request.

Taowyoo committed:
chore: add java example

Taowyoo committed:
ci: add CI for java-example

Taowyoo committed:
fix: add missing check on statement cert signature

xinyufort commented on key-attestation/java-example/.gitignore:
```diff
@@ -0,0 +1,19 @@
+# JDT-specific (Eclipse Java Development Tools)
+.classpath
+
+/.vscode
```
Nit of nits (please ignore if you're trying to merge this fast): you're missing a newline

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/Common.java:
```diff
@@ -0,0 +1,233 @@
+        if (!c.isEmpty()) {
+            builder.addRDN(BCStyle.C, c);
+        }
+        ;
```
Nit (please ignore if you're running low on time): why are there extra semicolons?

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/Common.java:
```diff
@@ -0,0 +1,233 @@
+    public static void isValidCrlUrl(String urlString) throws Exception {
+        URL url = new URL(urlString);
+        if (!url.getProtocol().equals("https")) {
+            throw new KeyAttestationStatementVerifyException("invalid CRL URL: invalid domain");
```
Did you switch the error messages from here and the if statement below?

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/Verify.java:
```diff
@@ -0,0 +1,225 @@
+    /**
+     * Verify given `attestationStatement`, `authorityChain` by using `trustRootCa`
+     *
+     * @param authorityChain       Certificate chain of all certificates expect
```
Nit (ignore if you're running low on time): I think you mean "except" instead of "expect"

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/Verify.java:
```diff
@@ -0,0 +1,225 @@
+            System.out.println("The signature in 'Fortanix DSM Key Attestation' certificate is valid.");
+        } catch (Exception e) {
+            throw new KeyAttestationStatementVerifyException(
+                    "The signature in 'Fortanix DSM Key Attestation' certificate is valid, " + e.toString());
```
Nit (feel free to ignore): I think you mean "invalid"

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/Verify.java:
```diff
@@ -0,0 +1,225 @@
+     * @param parent The certificate of issuer
+     * @throws Exception
+     */
+    public static void verify_cert_chain_signature(X509Certificate child, X509Certificate parent)
```
(Feel free to ignore)

I could be missing something, but where is this function being used?

Taowyoo replied:
updated to `verify_cert_signature`

Taowyoo committed:
refactor: add logging

Taowyoo committed:
build: remove unused dependencies

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/certchecker/FortanixRootCertChecker.java:
```diff
@@ -0,0 +1,95 @@
+        PublicKey rootPk = cert.getPublicKey();
+        if (rootPk instanceof RSAPublicKey) {
+            RSAPublicKey rootRsaPk = (RSAPublicKey) rootPk;
+            assert (rootRsaPk.getModulus().bitLength() >= 3072);
```
Nit (feel free to ignore): should we check that this is 4096?

(Also, should we have a constant to store these sort of #s?)

Taowyoo replied:
fixed

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/key_attestation_statement_verification/certchecker/KeyAttestationCaCertChecker.java:
```diff
@@ -0,0 +1,118 @@
+        PublicKey caPk = cert.getPublicKey();
+        if (caPk instanceof RSAPublicKey) {
+            RSAPublicKey caRsaPk = (RSAPublicKey) caPk;
+            assert (caRsaPk.getModulus().bitLength() >= 3072);
```
Same nit here (feel free to ignore): should this be 4096?

Taowyoo replied:
fixed

Taowyoo committed:
refactor: use better package name

xinyufort commented on key-attestation/java-example/pom.xml:
```diff
@@ -0,0 +1,75 @@
+      </plugin>
+    </plugins>
+  </reporting>
+</project>
```
Nit of nits (again, feel free to ignore): missing newline

Taowyoo replied:
fixed

Taowyoo committed:
feat: add functionality to decode and verify json data

xinyufort reviewed (commented):
part 1 review

Taowyoo committed:
chore: fix style and comments

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java:
```diff
@@ -0,0 +1,306 @@
+        String attestationStatementStr = keyAttestationResponse.getAttestationStatement().getStatement();
+        List<String> authorityChain = keyAttestationResponse.getAuthorityChain();
+        LOGGER.info(
+                String.format("Decoding Attestation Statement Certificate: %s", attestationStatementStr.toString()));
```
Nit (feel free to ignore): this log message is the same as the one a few lines down. Should we make them distinct (for better debugging)?

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java:
```diff
@@ -0,0 +1,306 @@
+        // it's parent
+        try {
+            verify_cert_chain_signature(authorityChain, trustRootCa, verifyCrl);
+            System.out.println("The signature in 'Fortanix DSM Key Attestation' certificate is invalid.");
```
Nit (feel free to ignore): don't you mean _valid_?

Taowyoo replied:
removed since this println is redundant

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java:
```diff
@@ -0,0 +1,306 @@
+                    "The signature in 'Fortanix DSM Key Attestation' certificate is invalid, " + e.toString());
+        }
+        LOGGER.info(String.format("Checking if '%s' certificate is correctly signed by '%s' certificate",
+                Common.DSM_CLUSTER_KEY_ATTESTATION_AUTHORITY_CN, Common.KEY_ATTESTATION_STATEMENT_CN));
```
Nit (feel free to ignore): wrong order?

Taowyoo replied:
Nice catch

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java:
```diff
@@ -0,0 +1,306 @@
+        if (authorityCertChainSize != Common.VALID_AUTHORITY_CERT_CHAIN_NUM) {
+            throw new KeyAttestationStatementVerifyException(
+                    String.format(
+                            "A valid authority certificates chain should contain %d certificate",
```
Nit (feel free to ignore): what if the chain is too long? E.g., we include extra certificates by accident...

My concern is whether this error message is the right thing to return, since I'm not sure if a cert chain of the wrong length necessarily implies that the chain does not contain the key attestation authority cert...

Taowyoo replied:
Since I assume the `authorityCertChain` comes from backend, so I also assume the length is three at current time point. We could update this part of code in future if we have any change.

xinyufort replied:
That's fair, I guess. Resolving for now.

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/certchecker/DsmKeyAttestationAuthorityCertChecker.java:
```diff
@@ -0,0 +1,155 @@
+                    "%s certificate 'Cluster node enrollment policy' extension policy item 'Node enrollment policy item: Minimum protection profile' should contain qualifier: 'Well-known protection profile: Fortanix FX2200' (%s), but get: %s",
+                    Common.DSM_CLUSTER_KEY_ATTESTATION_AUTHORITY_CN,
+                    Common.NODE_ENROLLMENT_POLICY_ITEM_WELL_KNOWN_PROTECTION_PROFILE_FORTANIX_FX2200_OID,
+                    policyItem1.getId()));
```
Did you mean `policyItem1Qualifier`?

Taowyoo replied:
nice catch

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/certchecker/DsmKeyAttestationAuthorityCertChecker.java:
```diff
@@ -0,0 +1,155 @@
+                    "%s certificate 'Cluster node enrollment policy' extension should contain policy item 'Node enrollment policy item: Site operator approval required' (%s), but get: %s",
+                    Common.DSM_CLUSTER_KEY_ATTESTATION_AUTHORITY_CN,
+                    Common.NODE_ENROLLMENT_POLICY_ITEM_SITE_OPERATOR_APPROVAL_REQUIRED_OID,
+                    policyItem1.getId()));
```
Did you mean `policyItem2`?

Taowyoo replied:
nice catch

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/certchecker/KeyAttestationStatementCertChecker.java:
```diff
@@ -0,0 +1,113 @@
+
+    private void check_public_key(X509Certificate cert) throws Exception {
+        PublicKey statementPk = cert.getPublicKey();
+        if (statementPk instanceof RSAPublicKey) {
```
How do you know these attestation statements are for RSA keys? What about EC keys?

Taowyoo replied:
You are right, since we cannot expect what kind of asymmetric key will be, I think we can skip the check of public key here.

xinyufort replied:
Resolving to reduce clutter then :)

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/types/asn1/KeyUsage.java:
```diff
@@ -0,0 +1,68 @@
+            throw new KeyAttestationStatementVerifyException(errStrPrefix + " invalid keyUsage extension");
+        }
+    }
+}
```
Nit of nits (ignore if you want): missing newline

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/types/asn1/NodeEnrollmentPolicyItem.java:
```diff
@@ -0,0 +1,85 @@
+
+        return sb.toString();
+    }
+}
```
Nit of nits (again, ignore if you want): missing newline

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/types/asn1/KeyUsage.java:
This is an interesting way of representing KeyUsage, but I guess it works.

(I was initially expecting something more like a bitflags or a boolean array, but I see that you've just defined an enum for the individual flags themselves, and are using regular boolean arrays for the actual key usage values...)

Taowyoo replied:
Yeah. As comments said, It is just a helper type.
So it only contains some minimal code to help check KeyUsage

xinyufort replied:
resolving to reduce clutter

Taowyoo replied:
Just refactor this class to be a bitflags like class

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/types/json/KeyAttestationStatement.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+}
```
Nit of nits (feel free to ignore): missing newline

xinyufort commented on key-attestation/java-example/src/test/resources/key-attestation-response.json:
```diff
@@ -0,0 +1,11 @@
+        "format": "x509_certificate",
+        "statement": "MIID/zCCAmegAwIBAgIUCVQHdgKkQdO34cZVHfE5cO1rQwUwDQYJKoZIhvcNAQELBQAwNjE0MDIGA1UEAwwrRm9ydGFuaXggRFNNIFNhYVMgS2V5IEF0dGVzdGF0aW9uIEF1dGhvcml0eTAgFw0yMzA5MDEwMDM2MDZaGA85OTk5MTIzMTIzNTk1OVowXjElMCMGA1UEAwwcRm9ydGFuaXggRFNNIEtleSBBdHRlc3RhdGlvbjE1MDMGCysGAQQBg4QaAQICDCRiMmI4N2Y4NC04MGQyLTRjNGMtYmE0MS03ZmExMmZiNGVmZjUwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDJnG/Pwu6ElZ7ySoMWw ...
+    }
+}
```
nit of nits: missing newline

Taowyoo replied:
got it

xinyufort commented on key-attestation/java-example/src/test/resources/key-attestation-statement.pem:
```diff
@@ -0,0 +1,129 @@
+yGypNrVAJI78/0/5zY+qkEivgh9xpnoAKHK4UT/cKhuLDCNQIEhB4fhnwhOuX1/1
+DFOz4Vdzshx2KL0eanblKF2i0ekR8tzv97a41mzM3Q/TXe0l9xVINzDynbePYHrs
+WBYyPcMCau5v4vFxobX+622kIrXllaU+gQ==
+-----END CERTIFICATE-----
```
nit of nits (feel free to ignore): missing newline

xinyufort reviewed (commented):
I've left another round of (mostly) nitpicks. Even if you choose to ignore my suggestions, I do hope you at least give my comments a read.

Otherwise, I don't see any red flags at all. (But hey, I'm not a Java pro, so :shrug:). Let me know when you want me to take another look and/or approve.

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/certchecker/KeyAttestationStatementCertChecker.java:
```diff
@@ -0,0 +1,113 @@
+        PublicKey statementPk = cert.getPublicKey();
+        if (statementPk instanceof RSAPublicKey) {
+            RSAPublicKey statementRsaPk = (RSAPublicKey) statementPk;
+            assert (statementRsaPk.getModulus().bitLength() >= 2048);
```
And even if the keys _are_ RSA, what makes you sure that they are 2K?

Taowyoo replied:
You are right, since we cannot expect what kind of asymmetric key will be, I think we can skip the check of public key here.

xinyufort replied:
resolving since you removed the check

Taowyoo committed:
style: fix some nit picks

Taowyoo committed:
fix: skip public key check in statement

Since we cannot expect what kind of asymmetric key will be in
Key Attestation statement, we can skip the check of public key for now.

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/Verify.java:
```diff
@@ -0,0 +1,305 @@
+        checkAuthorityChainLength(authorityChain);
+
+        LOGGER.info(String.format("Checking if '%s' certificate is correctly signed by '%s' certificate",
+                Common.DSM_CLUSTER_KEY_ATTESTATION_AUTHORITY_CN, Common.KEY_ATTESTATION_STATEMENT_CN));
```
I think the order is still incorrect here.

(This is a nitpick; feel free to ignore)

Taowyoo replied:
Oh, I see. Eagle eye :+1:

xinyufort reviewed (approved):
Looks good to me

Taowyoo committed:
refactor: refactor KeyUsage to KeyUsageExt

Refactor KeyUsageExt into a bitflags like type

Taowyoo committed:
fix: fix wrong logging

xinyufort commented on key-attestation/java-example/src/main/java/com/fortanix/keyattestationstatementverifier/types/asn1/KeyUsageExt.java:
Oh wow, you actually created a new KeyUsageExt type :astonished:

xinyufort reviewed (approved):
Thanks for writing this tool up!
